docs/sdg/sdg-flow-yaml.md

```diff
@@ -0,0 +1,101 @@
+
+## Problem Statement
+
+The `instructlab/sdg` library is introducing more extensive data generation pipelines. To enable customization, we should allow users of the library to provide a configuration file which defines a custom pipeline or extends an existing pipeline.
+
+In terms of the API constructs, a Pipeline is created from a sequence of “block configurations” which express how to instantiate and invoke the individual steps (aka blocks) in the pipeline. A Flow construct serves as a template from which a sequence of block configs can be generated.
+
+## Objective
+
+- Library users can specify a custom flow using a well-defined file format.
+- Library users can either use a custom flow standalone, or combine a custom flow with existing flows.
+- The file format and library can evolve substantially without making breaking changes.
+- Incompatible changes can be introduced while retaining support for existing custom flows for a deprecation period.
+
+## Proposal
+
+We will add an API that instantiates a `Flow` object from a file:
+
+```python
+flow = Flow.from_file('mycustomflow.yaml', params)
+block_configs = flow.render()
+```
+
+The YAML file format will mirror the API and look like this:
+
+```yaml
+block_configs:
+  - block_type: LLMBlock
+    block_config: # LLMBlock constructor kwargs
+      block_name: gen_knowledge
+      output_cols: ["question", "response"]
+    gen_kwargs: # kwargs for block.generate()
+      max_tokens": 2048,
+    drop_duplicates: ["question"]
+  - block_type: FilterByValueBlock
+    block_config:
+      block_name: filter_faithfulness
+      filter_column": judgment
+      filter_value: YES
+      operation: operator.eq
+      batch_kwargs:
+        num_procs: 8
+     drop_columns: ["judgment", "explanation"]
+```
+
+### Flow Params
+
+The following parameters are supplied to a Flow constructor and used by the render() method:
+
+- client - an OpenAI completions API client for talking to the teacher model via the serving backend (i.e.  llama-cpp or vLLM)
+- model_family - e.g. mixtral or merlinite
+- model_id - a path name for the specific teacher model being used
+- num_iters - how many samples to generate
+- batched - whether the model serving backend supports batched mode
+
+TODO: figure out how these should be supplied to the blocks, and whether any YAML templating based on these are necessary.
+
+### Model Prompts
+
+Based on whether model_family is mixtral or merlinite, a different prompt is used with the teacher model
+
+```python
+_MODEL_PROMPT_MIXTRAL = "<s> [INST] {prompt} [/INST]"
+_MODEL_PROMPT_MERLINITE = "'<|system|>\nYou are an AI language model developed by IBM Research. You are a cautious assistant. You carefully follow instructions. You are helpful and harmless and you follow ethical guidelines and promote positive behavior.\n<|user|>\n{prompt}\n<|assistant|>\n'"
+```
+
+TODO: figure out how to include the appropriate prompt in the block config.
```

Comment:
Maybe render() should be a method on the parent class, that calls something else implemented by the child class that does the stuff now, but also some validation and default setting.

For a block that doesn't have the model prompt set, we can set it based on either the explicit model family, or our best guess based on the model name (_get_model_family() does the logic here already).

Right now we are duplicating this logic everywhere, but really it could just be the default behavior unless specified otherwise.

Reply:
Only `LLMBlock` uses `model_prompt`, so I think `model_family` should be passed to `LLMBlock` and it will choose the correct prompt

At least for now ... if we find out later that we need to supply custom prompts, then I guess a custom flow author will need to provide prompts for each model serving backend